### DISCUSSION TITLE
disable sabbrev completions?

### DISCUSSION THREAD

**Discussion by mozirilla213** [akinomyoga/ble.sh#288] [Q&A] [answered]
Making abbreviations, makes them show in the menu completion as arguments for commands. `ble-sabbrev ab='ble-sabbrev' ; cd <TAB> # shows ab`. How could one disable this feature? for all commands and only for certain commands.

**Comment by akinomyoga** [comment]
ble.sh doesn't provide the configuration interface for this. If you would like to turn it off, you can rewrite the shell function `ble/complete/source:sabbrev`:

>  How could one disable this feature? for all commands

You can just replace the function with an empty command.

```bash
# blerc

function mozirilla213/suppress-sabbrev-completion.hook {
  function ble/complete/source:sabbrev { return 0; }
}
blehook/eval-after-load complete mozirilla213/suppress-sabbrev-completion.hook
```

> and only for certain commands.

This would be a bit tricky. One needs to rely on the ble.sh's internal functions. For example, to suppress the sabbrev completions for `echo`, you could do it in the following way:

```bash
# blerc

function mozirilla213/filter-sabbrev-completion.hook {
  ble/function#advice around ble/complete/source:sabbrev '
    local comp_words comp_line comp_point comp_cword
    ble/syntax:bash/extract-command "$_ble_edit_ind" || return 1
    if [[ ${comp_words[0]} != echo ]]; then
      ble/function#advice/do
    fi
  '
}
blehook/eval-after-load complete mozirilla213/filter-sabbrev-completion.hook
```

----

**Edit** (2023-03-09): An example implemention was added at [`blesh-contrib/config/github288-filter-sabbrev-completion.bash`](https://github.com/akinomyoga/blesh-contrib/blob/master/config/github288-filter-sabbrev-completion.bash).

```bash
ble-import config/github288-filter-sabbrev-completion

# Example: Always ignore sabbrev completions for command arguments
bleopt complete_source_sabbrev_command_ignore='*'

# Example: Ignore sabbrev completions for `echo` command and `ble/*` commands.
bleopt complete_source_sabbrev_command_ignore='echo:ble/*'
```

**Reply by mozirilla213** [reply]
Thanks! I tested these and work as expected, I'd mark it as answer unless you're planning making a configuration out of it

**Reply by akinomyoga** [reply]
You can mark this as an accepted answer, but I also added an example of implementing a new option `bleopt complete_source_sabbrev_command_ignore` in [`config/github288-filter-sabbrev-completion.bash`](https://github.com/akinomyoga/blesh-contrib/blob/master/config/github288-filter-sabbrev-completion.bash) in https://github.com/akinomyoga/blesh-contrib/commit/6c03906a6979a597ab5f9037721e45ec1917d246. The explanations can be found as code comments therein. I also added a link and usage examples in the above answer.

**Comment by akinomyoga** [comment]
I'm currently thinking about changing the ordering of the completions in the menu. Currently, the sabbrev words are shown first in the list, but I think that is the reason for the impression that these would feel annoying. I think the sabbrev words can be put at the end of the list.

**Reply by mozirilla213** [reply]
Yes indeed that was one impression of mine, but also the thing is that I made a lot of abbreviations to quickly type characters not available in my keyboard without changing the layout like I do in Vim with IPA symbols `ble-sabbrev "\r=ɾ" "\th=θ" "\sh=ʃ" # etc.`

So I did want a config to disable them altogether because otherwise on any tab completion on the start of a word my whole terminal is filled with them, always pushing my command line to the first line, and the previous output pushed back even further (And the completion words fill and cut the previous output when scrolling up the terminal which I think is a known limitation which is fine, but the point is that even on commands with very few completions, say `complete -W t t; t <TAB>` all my abbreviations are also shown)

**Reply by mozirilla213** [reply]
I guess this could be just fine if I just set `bleopt complete_menu_maxlines=8`. So yeah, ordering is the main annoying part.

**Reply by akinomyoga** [reply]
I have changed the ordering of candidates in commit a6f168d040a15dbd8b50faac97c65fe717c6fbb7.

> I made a lot of abbreviations to quickly type characters not available in my keyboard without changing the layout like I do in Vim with IPA symbols `ble-sabbrev "\r=ɾ" "\th=θ" "\sh=ʃ" # etc.`

I see. Maybe these are not currently so useful because the sabbrev expansions are always applied to the whole word (but not a substring in a word). I remembered that I also had an idea about a special type of sabbrev expansions that are activated even as a part of words. I named it *inline sabbrev* and decided to implement it. Commit 56208534664f4a48241d73d5c937d04cd51bdaaa supports it through `ble-sabbrev -i key=value`. You can now define these special symbols as

```bash
ble-sabbrev -i "\r=ɾ" "\th=θ" "\sh=ʃ"
```

so that it would expanded in any context such as 

```bash
$ echo 'abcde\th[press Space]
```

I also added an option `bleopt edit_magic_opts=inline-sabbrev-no-insert` in commit 56208534664f4a48241d73d5c937d04cd51bdaaa. This slightly changes the behavior of `magic-space` so that it doesn't insert a space when the inline sabbrev is performed. This could be useful when one wants to continue to input characters in a same word (i.e. without adding a space after the inline sabbrev).

**Reply by akinomyoga** [reply, marked as answer]
> So I did want a config to disable them altogether because otherwise on any tab completion on the start of a word my whole terminal is filled with them,

I also added options `bleopt complete_source_sabbrev_opts=no-empty-completion` and `bleopt complete_source_sabbrev_ignore=<pattern1>:<pattern2>:...` to adjust the detailed behavior of the sabbrev completion generation. The former makes the sabbrev completion generation inactive when the word to complete is an empty string. The latter specifies the exclude patterns for the sabbrev completions. I haven't yet added the description of these options in the wiki, but I'll later do that.

**Reply by akinomyoga** [reply]
> I haven't yet added the description of these options in the wiki, but I'll later do that.

I have added the wiki descriptions in [`bleopt complete_source_sabbrev_ignore`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A77-Completion#user-content-bleopt-complete_source_sabbrev_ignore) and [`bleopt complete_source_sabbrev_opts`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A77-Completion#user-content-bleopt-complete_source_sabbrev_opts).
